### PR TITLE
[FEAT] PWA

### PR DESCRIPTION
**Pull request by spencerdezartsmith**
# Description

This PR adds PWA (Progressive Web App) functionality to our app. The app will now have a service worker that will be handling pre caching, runtime caching and network requests. 

A player will now be able to "install" the game on their device. The game runs a background check for new releases on a 4 hourly cadence. If a new release is found then the device will install the new version and update the cache accordingly. The new service worker will take control of the clients when the app becomes "hidden" ie not visible to the player. 

There is an offline fallback page however the app will not function offline as all game actions still require the server.

This adds significant performance improvements as all static assets are cached and served Cache First. 

**EXCLUDES**: Install prompt (this will be manual for now)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

This has been tested using a production build which can be found [here](https://sunflower-land.com/pwa)

- [x] Safari (Mac, iOS)
- [x] Chrome (Mac, Android)
- [x] Brave (Mac)
- [x] Firefox (Mac)

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [ ] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/pwa.yml [hunk 1]
```diff
@@ -0,0 +1,107 @@
+name: PWA deploy
```
Comment (spencerdezartsmith) [inline .github/workflows/pwa.yml:1]: This was used to test this PWA in a production environment. You can check it out [here](https://sunflower-land.com/pwa)

Comment (spencerdezartsmith) [inline .github/workflows/pwa.yml:1]: This was used to test this PWA in a production environment. You can check it out [here](https://sunflower-land.com/pwa)

File: src/App.tsx [hunk 2]
```diff
@@ -8,21 +9,27 @@ import ErrorBoundary from "features/auth/components/ErrorBoundary";
 import { Navigation } from "./Navigation";
 import "./lib/i18n";
 import { WalletProvider } from "features/wallet/WalletProvider";
+import { useServiceWorkerUpdate } from "lib/utils/hooks/useServiceWorkerUpdate";
+
+const CHECK_FOR_UPDATE_INTERVAL = 1000 * 60 * 4;
```
Comment (spencerdezartsmith) [inline src/App.tsx:14]: I have set the interval to 4 hourly. Happy for feedback on other intervals though.

File: src/Navigation.tsx [hunk 3]
```diff
@@ -20,19 +20,14 @@ import { Builder } from "features/builder/Builder";
 import { AuthMachineState } from "features/auth/lib/authMachine";
 import { ZoomProvider } from "components/ZoomProvider";
 import { LoadingFallback } from "./LoadingFallback";
-
```
Comment (spencerdezartsmith) [inline src/Navigation.tsx:23]: This was removed due to a potential issue/bug when loading these lazily. On rare occasions these bundles are not loaded correctly. This presents a potential issue when caching static assets. I have removed for now as there are known issues with lazy loaded resources. 

The PWA performance gains have not really been impacted here.

File: src/Navigation.tsx [hunk 4]
```diff
@@ -71,9 +106,19 @@ export const Navigation: React.FC = () => {
 
   return (
     <>
-      <Auth />
+      <Auth showOfflineModal={showConnectionModal} />
       {showGame ? (
         <ZoomProvider>
+          <Modal show={showOrientationModal} centered backdrop={false}>
+            <Panel>
```
Comment (spencerdezartsmith) [inline src/Navigation.tsx:113]: Once PWA is up and running we will work on landscape mode for mobile.

File: src/components/ui/HudContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+import { createPortal } from "react-dom";
+
+/**
+ * Heads up display container which portals all Hud Components out to the body and applies safe area styling - should be used for all game Hud components
+ */
+export const HudContainer: React.FC = ({ children }) => {
```
Comment (spencerdezartsmith) [inline src/components/ui/HudContainer.tsx:7]: In order to make sure mobile safe areas are considered and handled correctly all hud's will now be wrapped in a `HudContainer`

File: src/components/ui/ReactPortal.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+function createWrapperAndAppendToBody(wrapperId: string) {
+  const wrapperElement = document.createElement("div");
+
+  wrapperElement.setAttribute("id", wrapperId);
+  document.body.appendChild(wrapperElement);
+
+  return wrapperElement;
+}
+
+export const ReactPortal: React.FC<Props> = ({
```
Comment (spencerdezartsmith) [inline src/components/ui/ReactPortal.tsx:17]: Reuseable component for using portals.

File: src/features/game/components/AudioMenu.tsx [hunk 7]
```diff
@@ -50,20 +50,23 @@ export const AudioMenu: React.FC<Props> = ({
...
 
   useEffect(() => {
-    if (musicPlayer.current) {
-      if (musicPaused) {
-        musicPlayer.current.pause();
-      } else {
-        musicPlayer.current.play();
-        musicPlayer.current.muted = false;
-      }
+    if (!musicPlayer.current) return;
```
Comment (spencerdezartsmith) [inline src/features/game/components/AudioMenu.tsx:53]: Handling a bug where `musicPlayer.current` was `undefined`.

File: vite.config.ts [hunk 8]
```diff
@@ -15,32 +15,78 @@ export default defineConfig({
...
+        globPatterns: ["**/*.{js,css,html}", "assets/*.{jpg,mp3,svg,gif,png}"],
+      },
+      filename: "sw.ts",
       manifest: {
+        name: "Sunflower Land",
+        id: "com.sunflower-land",
+        description:
+          "Plant, Chop, Mine, Craft & Collect at Sunflower Land. The MetaVerse game with endless resources.",
         short_name: "Sunflower Land",
+        start_url: "/pwa/",
```
Comment (spencerdezartsmith) [inline vite.config.ts:44]: NOTE: NEEDS TO BE CHANGED BEFORE MERGE
```suggestion
        start_url: "/",
```

File: src/lib/utils/hooks/useServiceWorkerUpdate.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
+/* eslint-disable no-console */
+import { useEffect, useRef } from "react";
+import { useRegisterSW } from "virtual:pwa-register/react";
+import lifecycle from "page-lifecycle/dist/lifecycle.mjs";
+
+const CHECK_FOR_UPDATE_INTERVAL = 1000 * 60 * 60 * 4; // 4 hours
```
Comment (spencerdezartsmith) [inline src/lib/utils/hooks/useServiceWorkerUpdate.ts:6]: Open to suggestions on the interval timing.

**Review by adamhannigan — APPROVED** [review]
One small step for Bumpkin, One giant leap for Sunflower Land!

**Pull request merged**